core/src/animation/transition.rs

```diff
@@ -67,40 +75,89 @@ pub trait TransitionState: Sized + 'static {
 pub trait Roc {
   /// Calc the rate of change of the duration from animation start.
   fn rate_of_change(&self, dur: Duration) -> AnimateProgress;
+
+  fn round_duration(&self) -> Duration;
+
+  /// Transition will apply with repeat times
+  fn repeat(self, repeat: f32) -> RepeatTransition<Self>
+  where
+    Self: Sized,
+  {
+    RepeatTransition { repeat, transition: self }
+  }
+
+  /// Transition will apply after the delay duration
+  fn delay(self, delay: Duration) -> DelayTransition<Self>
+  where
+    Self: Sized,
+  {
+    DelayTransition { delay, transition: self }
+  }
+
+  fn box_it(self) -> Box<dyn Roc>
+  where
+    Self: Sized + 'static,
+  {
+    Box::new(self)
+  }
 }
 
-impl<E: Easing> Roc for Transition<E> {
-  fn rate_of_change(&self, mut run_dur: Duration) -> AnimateProgress {
-    if self.duration.as_secs_f32().abs() < f32::EPSILON {
-      return AnimateProgress::Finish;
-    }
+pub trait RocBoxClone: Roc {
+  fn box_clone(&self) -> Box<dyn Roc>;
+}
 
-    if let Some(delay) = self.delay {
-      if run_dur < delay {
-        return AnimateProgress::Dismissed;
-      } else {
-        run_dur -= delay;
-      }
+impl<T: Roc + Clone + 'static> RocBoxClone for T {
+  fn box_clone(&self) -> Box<dyn Roc> { self.clone().box_it() }
+}
+
+impl<T: Roc> Roc for DelayTransition<T> {
+  fn rate_of_change(&self, dur: Duration) -> AnimateProgress {
+    if dur < self.delay {
+      return AnimateProgress::Dismissed;
     }
+    self.transition.rate_of_change(dur - self.delay)
+  }
 
-    let repeat = self.repeat.unwrap_or(1.);
+  fn round_duration(&self) -> Duration { self.delay + self.transition.round_duration() }
+}
 
-    let rounds = run_dur.as_secs_f32() / self.duration.as_secs_f32();
+impl<T: Roc> Roc for RepeatTransition<T> {
+  fn rate_of_change(&self, dur: Duration) -> AnimateProgress {
+    let repeat = self.repeat;
+    let duration = self.transition.round_duration();
+    let rounds = dur.as_secs_f32() / duration.as_secs_f32();
     if rounds > repeat {
       return AnimateProgress::Finish;
     }
+    let rate = match self
+      .transition
+      .rate_of_change(duration.mul_f32(rounds.fract()))
+    {
+      AnimateProgress::Dismissed => 0.,
+      AnimateProgress::Finish => 1.,
+      AnimateProgress::Between(p) => p,
+    };
+    AnimateProgress::Between(rate)
+  }
 
-    let time_rate = run_dur.as_secs_f32() / self.duration.as_secs_f32() - rounds.floor();
+  fn round_duration(&self) -> Duration {
```

Comment:
We should not add `round_duration` in `Roc` only for `RepeatTransition`, As an alternative, we should include the time exceeded when the `rate_of_change` returns the completion state -- `AnimateProgress::FinishOver(Duration)`,